cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerService.java

```diff
@@ -42,45 +55,106 @@
 public class TaskWorkerService extends AbstractIdleService {
 
   private static final Logger LOG = LoggerFactory.getLogger(TaskWorkerService.class);
+  private static final Gson GSON = new Gson();
 
+  private final CConfiguration cConf;
   private final DiscoveryService discoveryService;
   private final NettyHttpService httpService;
+  private final ArtifactManagerFactory artifactManagerFactory;
+  private final RunnableTaskLauncher taskLauncher;
   private Cancellable cancelDiscovery;
   private InetSocketAddress bindAddress;
 
   @Inject
   TaskWorkerService(CConfiguration cConf,
                     SConfiguration sConf,
-                    DiscoveryService discoveryService) {
+                    DiscoveryService discoveryService,
+                    ArtifactManagerFactory artifactManagerFactory) {
+    this.cConf = cConf;
     this.discoveryService = discoveryService;
+    this.artifactManagerFactory = artifactManagerFactory;
+    this.taskLauncher = new RunnableTaskLauncher(cConf);
 
     NettyHttpService.Builder builder = new CommonNettyHttpServiceBuilder(cConf, Constants.Service.TASK_WORKER)
       .setHost(cConf.get(Constants.TaskWorker.ADDRESS))
       .setPort(cConf.getInt(Constants.TaskWorker.PORT))
       .setExecThreadPoolSize(cConf.getInt(Constants.TaskWorker.EXEC_THREADS))
       .setBossThreadPoolSize(cConf.getInt(Constants.TaskWorker.BOSS_THREADS))
       .setWorkerThreadPoolSize(cConf.getInt(Constants.TaskWorker.WORKER_THREADS))
+      .setChannelPipelineModifier(new ChannelPipelineModifier() {
+        @Override
+        public void modify(ChannelPipeline pipeline) {
+          pipeline.addAfter("compressor", "decompressor", new HttpContentDecompressor());
+        }
+      })
       .setHttpHandlers(new TaskWorkerHttpHandlerInternal(cConf, this::stopService));
 
     if (cConf.getBoolean(Constants.Security.SSL.INTERNAL_ENABLED)) {
       new HttpsEnabler().configureKeyStore(cConf, sConf).enable(builder);
     }
-    httpService = builder.build();
+    this.httpService = builder.build();
   }
 
   @Override
   protected void startUp() throws Exception {
-    LOG.info("Starting TaskWorkerService");
+    preloadArtifacts();
+
+    LOG.debug("Starting TaskWorkerService");
     httpService.start();
     bindAddress = httpService.getBindAddress();
     cancelDiscovery = discoveryService.register(
       ResolvingDiscoverable.of(URIScheme.createDiscoverable(Constants.Service.TASK_WORKER, httpService)));
-    LOG.info("Starting TaskWorkerService has completed");
+    LOG.debug("Starting TaskWorkerService has completed");
+  }
+
+  /**
+   * Preloading artifacts by running a {@link SystemAppTask} for each of the artifact.
+   */
+  private void preloadArtifacts() {
+    Set<String> artifacts = new HashSet<>(cConf.getTrimmedStringCollection(Constants.TaskWorker.PRELOAD_ARTIFACTS));
```

Comment:
I am having trouble finding where the value for this is being set. cdap-default is having an empty value. Is it set while loading system artifacts?

Reply:
Default is nothing to preload. The value can be overridden by CDAP-site.xml, which is generated via CDAP yaml in k8s environment